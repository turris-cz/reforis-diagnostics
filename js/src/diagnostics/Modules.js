/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import {
    API_STATE,
    useAPIGet,
    useForm,
    withErrorMessage,
    withSpinner,
} from "foris";

import API_URLs from "API";
import ModulesForm from "./ModulesForm";

Modules.propTypes = {
    onReload: PropTypes.func.isRequired,
};

export default function Modules({ onReload }) {
    const [getModulesResponse, getModules] = useAPIGet(API_URLs.modules);
    useEffect(() => {
        getModules();
    }, [getModules]);

    const [formState, setFormValue] = useForm(bypassValidator);
    const formData = formState.data || {};
    useEffect(() => {
        if (getModulesResponse.state === API_STATE.SUCCESS) {
            const modules = getModulesResponse.data.modules.reduce(
                (modulesAccumulator, module) => ({
                    ...modulesAccumulator,
                    [module.module_id]: true,
                }),
                {}
            );
            setFormValue((value) => ({ $set: { modules: value } }))({
                target: { value: modules },
            });
        }
    }, [getModulesResponse, setFormValue]);
    const descriptions =
        getModulesResponse.data &&
        getModulesResponse.data.modules.reduce(
            (descriptionsAccumulator, module) => ({
                ...descriptionsAccumulator,
                [module.module_id]: module.description,
            }),
            {}
        );
    return (
        <>
            <h2>{_("Modules")}</h2>
            <ModulesFormWithErrorAndSpinner
                apiState={getModulesResponse.state}
                onReload={onReload}
                modules={formData.modules}
                descriptions={descriptions}
                setFormValue={setFormValue}
            />
        </>
    );
}

// Has to be defined outside component as it would cause re-renders
function bypassValidator() {
    return {};
}

const withSpinnerOnModules = withSpinner((props) => !props.modules);
const ModulesFormWithErrorAndSpinner = withErrorMessage(
    withSpinnerOnModules(ModulesForm)
);
