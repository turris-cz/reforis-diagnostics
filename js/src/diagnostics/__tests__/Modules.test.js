/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, waitForElement, getByText, getByLabelText, fireEvent, wait } from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from 'jest-mock-axios';

import Modules from "../Modules";

describe("<Modules />", () => {
    let componentContainer;
    const modules = ["foo_module", "bar_module", "fizz_module", "buzz_module"];
    const handleReload = jest.fn();

    beforeEach(() => {
        const { container } = render(<Modules onReload={handleReload} />);
        componentContainer = container;
    });

    it("should display spinner", async () => {
        expect(componentContainer).toMatchSnapshot();
    });

    it("should render modules", async () => {
        expect(mockAxios.get).toBeCalledWith("/reforis/diagnostics/api/modules", expect.anything());
        mockAxios.mockResponse({data: {modules: modules}});
        await waitForElement(() => getByText(componentContainer, "Modules"));
        expect(componentContainer).toMatchSnapshot();
    });

    it("should handle error on fetching modules", async () => {
        expect(mockAxios.get).toBeCalledWith("/reforis/diagnostics/api/modules", expect.anything());
        mockJSONError();
        await waitForElement(() => getByText(componentContainer, "Modules"));
        expect(componentContainer).toMatchSnapshot();
    });

    it("should select all modules", async () => {
        mockAxios.mockResponse({data: {modules: modules}});
        await waitForElement(() => getByText(componentContainer, "Modules"));

        // Default state - all unselected and button disabled
        expect(getByText(componentContainer, "Generate report").disabled).toBe(true);
        modules.forEach((module) => {
            expect(getByLabelText(componentContainer, module).checked).toBe(false);
        });

        // All selected, button enabled
        fireEvent.click(getByText(componentContainer, "Select all"));
        expect(getByText(componentContainer, "Generate report").disabled).toBe(false);
        modules.forEach((module) => {
            expect(getByLabelText(componentContainer, module).checked).toBe(true);
        });
    });

    it("should generate report", async () => {
        mockAxios.mockResponse({data: {modules: modules}});
        await waitForElement(() => getByText(componentContainer, "Modules"));

        fireEvent.click(getByText(componentContainer, modules[0]));
        fireEvent.click(getByText(componentContainer, "Generate report"));
        expect(mockAxios.post).toBeCalledWith(
            "/reforis/diagnostics/api/reports", {modules: [modules[0]]}, expect.anything()
        );

        // Response to POST report
        mockAxios.mockResponse({data: {}});
        await waitForElement(() => getByText(componentContainer, "Modules"));
        expect(handleReload).toHaveBeenCalledTimes(1);
    });

    it("should handle error on generating report", async () => {
        mockAxios.mockResponse({data: {modules: modules}});
        await waitForElement(() => getByText(componentContainer, "Modules"));

        fireEvent.click(getByText(componentContainer, modules[0]));
        fireEvent.click(getByText(componentContainer, "Generate report"));
        expect(mockAxios.post).toBeCalledWith(
            "/reforis/diagnostics/api/reports", {modules: [modules[0]]}, expect.anything()
        );

        // Response to POST report
        mockJSONError();
        await wait(() => {
            expect(mockSetAlert).toHaveBeenCalledWith("Cannot generate report");
        });
    });
});
