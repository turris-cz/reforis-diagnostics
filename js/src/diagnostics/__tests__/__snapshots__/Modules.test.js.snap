// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modules /> should display spinner 1`] = `
<div>
  <h3>
    Modules
  </h3>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;

exports[`<Modules /> should handle error on fetching modules 1`] = `
<div>
  <h3>
    Modules
  </h3>
  <p
    class="text-center text-danger"
  >
    An error occurred while fetching data.
  </p>
</div>
`;

exports[`<Modules /> should render modules 1`] = `
<div>
  <h3>
    Modules
  </h3>
  <p>
    Select modules which you want to include in the report.
  </p>
  <form
    class="col-sm-12 offset-lg-1 col-lg-10 p-0 mb-3"
  >
    <div
      style="font-weight: bold;"
    >
      <div
        class="form-group"
      >
        <div
          class="custom-control custom-checkbox "
        >
          <input
            checked=""
            class="custom-control-input"
            id="1"
            type="checkbox"
          />
          <label
            class="custom-control-label"
            for="1"
          >
            Select all
          </label>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-sm-12 col-md-6 col-lg-4"
      >
        <div
          class="form-group"
        >
          <div
            class="custom-control custom-checkbox "
          >
            <input
              checked=""
              class="custom-control-input"
              id="2"
              type="checkbox"
            />
            <label
              class="custom-control-label"
              for="2"
            >
              foo_module
              <small
                class="form-text text-muted"
              >
                Boo module description.
              </small>
            </label>
          </div>
        </div>
      </div>
      <div
        class="col-sm-12 col-md-6 col-lg-4"
      >
        <div
          class="form-group"
        >
          <div
            class="custom-control custom-checkbox "
          >
            <input
              checked=""
              class="custom-control-input"
              id="3"
              type="checkbox"
            />
            <label
              class="custom-control-label"
              for="3"
            >
              bar_module
              <small
                class="form-text text-muted"
              >
                Bar module description.
              </small>
            </label>
          </div>
        </div>
      </div>
      <div
        class="col-sm-12 col-md-6 col-lg-4"
      >
        <div
          class="form-group"
        >
          <div
            class="custom-control custom-checkbox "
          >
            <input
              checked=""
              class="custom-control-input"
              id="4"
              type="checkbox"
            />
            <label
              class="custom-control-label"
              for="4"
            >
              fizz_module
              <small
                class="form-text text-muted"
              >
                Fizz module description.
              </small>
            </label>
          </div>
        </div>
      </div>
      <div
        class="col-sm-12 col-md-6 col-lg-4"
      >
        <div
          class="form-group"
        >
          <div
            class="custom-control custom-checkbox "
          >
            <input
              checked=""
              class="custom-control-input"
              id="5"
              type="checkbox"
            />
            <label
              class="custom-control-label"
              for="5"
            >
              buzz_module
              <small
                class="form-text text-muted"
              >
                Buzz module description.
              </small>
            </label>
          </div>
        </div>
      </div>
    </div>
    <div
      class="text-right"
    >
      <button
        class="btn btn-primary  col-sm-12 col-lg-3"
        type="submit"
      >
         
         
        Generate report
      </button>
    </div>
  </form>
</div>
`;
